UnitsNet/QuantityValue.cs

```diff
@@ -72,28 +83,84 @@ private QuantityValue(decimal val)
 
         /// <summary>Explicit cast from <see cref="QuantityValue"/> to <see cref="double"/>.</summary>
         public static explicit operator double(QuantityValue number)
-        {
-            // double -> decimal -> zero (since we can't implement the default struct ctor)
-            return number._value ?? (double) number._valueDecimal.GetValueOrDefault();
-        }
+            => number.Type switch
+            {
+                UnderlyingDataType.Decimal => (double)number._decimalValue,
+                UnderlyingDataType.Double => number._doubleValue,
+                _ => throw new System.NotImplementedException()
+            };
 
         #endregion
 
         #region To decimal
 
         /// <summary>Explicit cast from <see cref="QuantityValue"/> to <see cref="decimal"/>.</summary>
         public static explicit operator decimal(QuantityValue number)
-        {
-            // decimal -> double -> zero (since we can't implement the default struct ctor)
-            return number._valueDecimal ?? (decimal) number._value.GetValueOrDefault();
-        }
+            => number.Type switch
+            {
+                UnderlyingDataType.Decimal => number._decimalValue,
+                UnderlyingDataType.Double => (decimal)number._doubleValue,
+                _ => throw new System.NotImplementedException()
+            };
 
         #endregion
 
         /// <summary>Returns the string representation of the numeric value.</summary>
         public override string ToString()
+            => Type switch
+            {
+                UnderlyingDataType.Decimal => _decimalValue.ToString(),
+                UnderlyingDataType.Double => _doubleValue.ToString(),
+                _ => throw new System.NotImplementedException()
+            };
+
+        /// <summary>
+        ///     Describes the underlying type of a <see cref="QuantityValue"/>.
+        /// </summary>
+        public enum UnderlyingDataType : byte
         {
-            return _value.HasValue ? _value.ToString() : _valueDecimal.ToString();
+            /// <inheritdoc cref="decimal"/>
+            /// <remarks>Has to be 0 due to the bit structure of <see cref="decimal"/>.</remarks>
+            Decimal = 0,
+            /// <inheritdoc cref="double"/>
+            Double = 1
         }
+
+        #region just for debugging
+#if DEBUG
+        [FieldOffset(0)]
```

Comment:
You might be able to use this instead
 https://docs.microsoft.com/en-us/dotnet/framework/debug-trace-profile/enhancing-debugging-with-the-debugger-display-attributes